DSpace/dspace-angular
# Authorization for downloads of restricted bitstreams

## References
This PR fixes #631
This PR requires REST API PR https://github.com/DSpace/DSpace/pull/2783

## Description
This PR adds support for authorization of downloading restricted bitstreams through regular links.

## Instructions for Reviewers
Changes made:
- Added `TokenResponseParsingService` along with a `TokenPostRequest` and `TokenResponse` for requesting and retrieving [short-lived tokens](https://github.com/DSpace/Rest7Contract/blob/master/authentication.md#request-short-lived-token) from the REST API.
- Added `getShortlivedToken()` to auth-request-service, creating the post request and returning the token from the response
- Added `getShortlivedToken()` to auth-service, checking if the user is authenticated, returning the token if he is, null if he isn’t
- Modified `downloadFile()` from FileService to retrieve the token, add it to the url (when not empty) and set window.location.href to the new url
- Created new `FileDownloadLinkComponent`, rendering a `<a>` element, containing a click event if the user is authenticated. The click event calls the file service's downloadFile method and returns false (to still allow hovering over the link to display the original href)
- Replaced regular download links of bitstreams with this new component
- Implemented tests and added JSDocs where necessary

Warning:
Reviewers may be affected by issues #711 and #712 

How to test:
- Visit an item's simple and/or full item page. If the REST API supports short-lived tokens and a user is authenticated, it should add a token to the download, allowing access to restricted bitstreams.
**Note**: Only publication simple item pages contain bitstream links, so make sure to visit a publication.

## Checklist
_This checklist provides a reminder of what we are going to look for when reviewing your PR. You need not complete this checklist prior to creating your PR (draft PRs are always welcome). If you are unsure about an item in the checklist, don't hesitate to ask. We're here to help!_
    * Include tests for different user types (if behavior differs), including: (1) Anonymous user, (2) Logged in user (non-admin), and (3) Administrator.
    * Include tests for error scenarios, e.g. when errors/warnings should appear (or buttons should be disabled).
    * For bug fixes, include a test that reproduces the bug and proves it is fixed. For clarity, it may be useful to provide the test in a separate commit from the bug fix.
